amgut/static/js/interactive.js

```diff
@@ -0,0 +1,292 @@
+      var thisArg = arguments[1];
+      var value;
+
+      for (var i = 0; i < length; i++) {
+        value = list[i];
+        if (predicate.call(thisArg, value, i, list)) {
+          return i;
+        }
+      }
+      return -1;
+    };
+  }
+});
+
+function group(list, prop) {
+  var grouped = {};
+  for(var i=0;i<list.length;i++) {
+    var key = list[i][prop];
+    grouped[key] = grouped[key] || [];
+    grouped[key].push(list[i]);
+  }
+  return grouped;
+}
+
+function getAvg(list) {
+  return list.reduce(function (x, y) {
+    return x + y;
+  }) / list.length;
+}
+
+function sumArrays(ar1, ar2) {
+  var sum = [];
+  if (ar1.length == ar2.length) {
+      for (var i=0; i<ar1.length;i++) {
+          sum.push(ar1[i] + ar2[i]);
+      }
+  }
+  else {
+    throw "Arrays unequal length: " + ar1.length + "  " + ar2.length;
+  }
+  return sum;
+}
+
+function findOtu(list, otu, level) {
+  var lvl = level || 'label';
+  function inner(element, index, array) {
+    if(element[lvl] == this) { return true; }
+    return false;
+  }
+  return list.findIndex(inner, otu);
+}
+
+function filterSites(list, sites) {
+
+}
+
+function collapse(dataset, level, max, prev_level, focus, sites) {
+  var phylum_colors = {
+    'Firmicutes': ['pink', 'darkred'],
+    'Bacteroidetes': ['#FFD27F', '#BE5E00'],
+    'Proteobacteria': ['lightyellow', 'yellow'],
+    'Actinobacteria': ['#F3DBAB', '#735B2B'],
+    'Verrucomicrobia': ['#C5D991', '#465A12'],
+    'Tenericutes': ['lightgreen', 'darkgreen'],
+    'Cyanobacteria': ['lightcyan', 'darkcyan'],
+    'Fusobacteria': ['lightblue', 'darkblue']
+  };
+  var phylum_order = ['Firmicutes', 'Bacteroidetes', 'Proteobacteria', 'Actinobacteria', 'Verrucomicrobia', 'Tenericutes', 'Cyanobacteria', 'Fusobacteria'];
+
+  var max_otus = max || 8;
+  var focus_taxa = focus || null;
+  var filter_to = sites || null;
+  var groups = group(dataset, level);
+  var collapsed = {};
+  var summaries = [];
+  for(var g in groups) {
+    if(focus !== null && groups[g][0][prev_level] != focus_taxa) { continue; }
+    var otu = { label: g,  data: groups[g][0].data, phylum: groups[g][0].phylum};
+    if(g === "") { otu.label = 'Unspecified'; }
+    // Collapse the OTUs found in the group into a single
+    for(var i=1;i<groups[g].length;i++) {
+      otu.data = sumArrays(otu.data, groups[g][i].data);
+    }
+    // Round to four decimals
+    for(var i=0;i<otu.data.length;i++) { otu.data[i] = Math.floor(otu.data[i] * 10000) / 10000; }
+
+    collapsed[g] = otu;
+    summaries.push([g, getAvg(otu.data)]);
+  }
+
+  //collapse further to max number of OTUs and color properly
+  // Making sure to keep phyla order
+  var phyla_dict = {};
+  sorted = summaries.sort(function(a, b) { return b[1] - a[1]; });
+  var loop = sorted.length;
+  if(loop > max) { loop = max; }
+  for(var i=0;i<loop;i++) {
+    // Add to existing phylum list or create new if not exist
+    if(phyla_dict.hasOwnProperty(collapsed[sorted[i][0]].phylum)) {
```

Comment:
the code in this if-statement could be simplified by doing the complex indexing for `collapsed[sorted[i][0]]` a single time